File: salazar/salazar/agents/planner.py
"""Planner agent — decomposes app_spec.md into feature_list.json."""

import logging
from pathlib import Path

from claude_agent_sdk import query, ResultMessage

from salazar.client import make_options, OUTPUT_DIR

logger = logging.getLogger("harness.planner")

PROMPT_PATH = Path(__file__).resolve().parent.parent / "prompts" / "planner.md"


async def run_planner(app_spec_path: Path) -> bool:
    """Run the planner agent to create feature_list.json.

    Args:
        app_spec_path: Path to the app_spec.md file.

    Returns:
        True if feature_list.json was created successfully.
    """
    logger.info("[planner] Starting planner agent")

    # Read the system prompt and app spec
    system_prompt = PROMPT_PATH.read_text()
    app_spec = app_spec_path.read_text()

    # Copy app_spec into output directory
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    (OUTPUT_DIR / "app_spec.md").write_text(app_spec)

    prompt = (
        f"Here is the product specification. Decompose it into a comprehensive "
        f"feature list with BDD scenarios.\n\n"
        f"---\n\n{app_spec}"
    )

    options = make_options(
        system_prompt=system_prompt,
        role="planner",
        max_budget_usd=10.0,  # Planner is a one-shot — shouldn't cost much
    )

    import asyncio

    async def _run():
        nonlocal result_text
        async for message in query(prompt=prompt, options=options):
            if isinstance(message, ResultMessage):
                cost = message.total_cost_usd or 0.0
                logger.info(f"[planner] Session complete: cost=${cost:.4f}, turns={message.num_turns}")
                result_text = f"cost={cost:.4f}"

    result_text = ""
    try:
        await asyncio.wait_for(_run(), timeout=600)  # 10 min timeout
    except asyncio.TimeoutError:
        logger.error("[planner] Session timed out after 10 minutes")

    # Verify output
    feature_list = OUTPUT_DIR / "feature_list.json"
    if feature_list.exists():
        logger.info(f"[planner] Created feature_list.json ({feature_list.stat().st_size} bytes)")
        return True
    else:
        logger.error("[planner] feature_list.json was not created")
        return False

File: src/engine/agents/planner.ts
/**
 * Planner agent — decomposes app_spec.md into feature_list.json.
 *
 * One-shot Claude Code session that reads a product spec and writes
 * feature_list.json to the output directory.
 */

import { readFileSync, copyFileSync, existsSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { query } from "@anthropic-ai/claude-agent-sdk";
import { makeQueryOptions } from "../client.js";
import type { SalazarConfig } from "../../lib/types.js";

export interface PlannerResult {
  success: boolean;
  costUsd: number;
}

export async function runPlanner(
  specPath: string,
  outputDir: string,
  config: SalazarConfig
): Promise<PlannerResult> {
  console.log("[planner] Starting planner agent");

  const specText = readFileSync(specPath, "utf-8");
  mkdirSync(outputDir, { recursive: true });
  const destSpecPath = join(outputDir, "app_spec.md");
  copyFileSync(specPath, destSpecPath);
  console.log(`[planner] Copied spec to ${destSpecPath}`);

  const featureListPath = join(outputDir, "feature_list.json");

  const prompt =
    `Here is the product specification. Decompose it into a comprehensive ` +
    `feature list with BDD scenarios.\n\n` +
    `IMPORTANT: Write the output to this exact absolute path: ${featureListPath}\n` +
    `Also read the spec from: ${destSpecPath}\n\n` +
    `---\n\n${specText}`;

  const options = makeQueryOptions({
    role: "planner",
    config,
    cwd: outputDir,
    maxTurns: 50,
    maxBudgetUsd: 10.0,
  });

  let costUsd = 0;

  try {
    for await (const message of query({ prompt, options })) {
      if ("result" in message) {
        // Session completed — result message
        // Cost may be available on the message depending on SDK version;
        // cast to any to probe without hard-coding a potentially-wrong type.
        const result = message as Record<string, unknown>;
        if (typeof result["total_cost_usd"] === "number") {
          costUsd = result["total_cost_usd"];
        } else if (typeof result["cost_usd"] === "number") {
          costUsd = result["cost_usd"];
        }
        console.log(`[planner] Session complete: cost=$${costUsd.toFixed(4)}`);
      }
    }
  } catch (err) {
    console.error("[planner] Session error:", err);
    return { success: false, costUsd };
  }

  // Verify output
  if (existsSync(featureListPath)) {
    const { statSync } = await import("node:fs");
    const size = statSync(featureListPath).size;
    console.log(`[planner] Created feature_list.json (${size} bytes)`);
    return { success: true, costUsd };
  } else {
    console.error("[planner] feature_list.json was not created");
    return { success: false, costUsd };
  }
}
